### PR TITLE
[language, types] Clean up dead, untested code in language

### LINKED ISSUE #4220: [Bug] 81 unused, untested members in `language/`
# 🐛 Bug

I found 81 unused, untested members in `language/`. They are listed, with a proof of concept removal diff that passes all tests, here:
https://github.com/libra/libra/pull/4148 (x 72)
https://github.com/libra/libra/pull/4176 (x 9)

### PR DESCRIPTION
**Pull request by huitseeker**
I know, I know, @dariorussi doesn't like this one bit, and I suspect he might not be the only one. But I figured that as development speed changes with code freezes, etc, this would be as good a time as any to tackle the cruft.

And no other place in the code base has anywhere near that amount of dead code.

Individualized commits so I can roll back stuff piecemeal if you tell me you want to keep it in review.

Fixes #4220

### CODE REVIEW

File: language/move-core/types/src/identifier.rs [hunk 1]
```diff
@@ -167,11 +161,6 @@ impl IdentStr {
         is_valid(s.as_ref())
     }
 
-    /// Returns the length of `self` in bytes.
-    pub fn len(&self) -> usize {
```
Comment (tnowacki) [inline language/move-core/types/src/identifier.rs:171]: Please revert

File: language/compiler/bytecode-source-map/src/marking.rs [hunk 2]
```diff
@@ -92,52 +90,4 @@ impl MarkedSourceMapping {
             struct_marks: HashMap::new(),
         }
     }
-
-    pub fn mark_code_offset(
```
Comment (tnowacki) [inline language/compiler/bytecode-source-map/src/marking.rs:96]: Please revert all to the bytecode source map

File: language/borrow-graph/src/graph.rs [hunk 3]
```diff
@@ -90,18 +90,6 @@ impl<Loc: Copy, Lbl: Clone + Ord> BorrowGraph<Loc, Lbl> {
         returned_edges
     }
 
-    /// Return the incoming edges into id
-    pub fn in_edges(&self, id: RefID) -> Vec<(Loc, RefID, Path<Lbl>, bool)> {
```
Comment (tnowacki) [inline language/borrow-graph/src/graph.rs:94]: Please revert

File: language/move-core/types/src/value.rs [hunk 4]
```diff
@@ -61,13 +61,6 @@ impl MoveKind {
         }
     }
 
-    pub fn is_copyable(&self) -> bool {
```
Comment (tnowacki) [inline language/move-core/types/src/value.rs:64]: Please revert

File: language/move-core/types/src/value.rs [hunk 5]
```diff
@@ -82,10 +75,6 @@ impl MoveKindInfo {
         self.kind().is_resource()
     }
 
-    pub fn is_copyable(&self) -> bool {
```
Comment (tnowacki) [inline language/move-core/types/src/value.rs:85]: Please revert

File: language/move-core/types/src/value.rs [hunk 6]
```diff
@@ -119,10 +108,6 @@ impl MoveStruct {
     pub fn fields(&self) -> &[MoveValue] {
         &self.0
     }
-
-    pub fn into_inner(self) -> Vec<MoveValue> {
```
Comment (tnowacki) [inline language/move-core/types/src/value.rs:123]: Please revert

File: language/move-lang/src/cfgir/cfg.rs [hunk 7]
```diff
@@ -118,20 +118,6 @@ impl<'a> BlockCFG<'a> {
     pub fn block(&self, label: Label) -> &BasicBlock {
         self.blocks.get(&label).unwrap()
     }
-
-    pub fn block_mut(&mut self, label: Label) -> &mut BasicBlock {
```
Comment (tnowacki) [inline language/move-lang/src/cfgir/cfg.rs:122]: Please revert

File: language/move-lang/src/compiled_unit.rs [hunk 8]
```diff
@@ -70,15 +70,6 @@ impl CompiledUnit {
         serialized
     }
 
-    #[allow(dead_code)]
-    pub fn serialize_debug(self) -> Vec<u8> {
```
Comment (tnowacki) [inline language/move-lang/src/compiled_unit.rs:74]: Please revert

File: language/move-lang/src/expansion/ast.rs [hunk 9]
```diff
@@ -297,10 +297,6 @@ impl SpecId {
     pub fn new(u: usize) -> Self {
         SpecId(u)
     }
-
-    pub fn inner(self) -> usize {
```
Comment (tnowacki) [inline language/move-lang/src/expansion/ast.rs:301]: Please revert

File: language/move-lang/src/hlir/ast.rs [hunk 10]
```diff
@@ -397,22 +397,6 @@ impl BaseType_ {
     pub fn bool(loc: Loc) -> BaseType {
         Self::builtin(loc, BuiltinTypeName_::Bool, vec![])
     }
-
-    pub fn address(loc: Loc) -> BaseType {
```
Comment (tnowacki) [inline language/move-lang/src/hlir/ast.rs:401]: Please revert

File: language/move-lang/src/hlir/ast.rs [hunk 11]
```diff
@@ -424,22 +408,6 @@ impl SingleType_ {
         Self::base(BaseType_::bool(loc))
     }
 
-    pub fn address(loc: Loc) -> SingleType {
```
Comment (tnowacki) [inline language/move-lang/src/hlir/ast.rs:427]: Please revert

File: language/move-lang/src/hlir/ast.rs [hunk 12]
```diff
@@ -461,22 +429,6 @@ impl Type_ {
         Self::single(SingleType_::bool(loc))
     }
 
-    pub fn address(loc: Loc) -> Type {
```
Comment (tnowacki) [inline language/move-lang/src/hlir/ast.rs:464]: Please revert

File: language/move-lang/src/naming/ast.rs [hunk 13]
```diff
@@ -377,19 +377,6 @@ impl BuiltinFunction_ {
             _ => None,
         }
     }
-
-    pub fn display_name(&self) -> &'static str {
```
Comment (tnowacki) [inline language/move-lang/src/naming/ast.rs:381]: Please revert

File: language/move-lang/src/parser/ast.rs [hunk 14]
```diff
@@ -601,12 +601,6 @@ impl Kind_ {
     }
 }
 
-impl Type_ {
-    pub fn unit(loc: Loc) -> Type {
```
Comment (tnowacki) [inline language/move-lang/src/parser/ast.rs:605]: Please revert

File: language/move-lang/src/shared/remembering_unique_map.rs [hunk 15]
```diff
@@ -28,36 +28,15 @@ impl<K: TName, V> RememberingUniqueMap<K, V> {
         self.map.is_empty()
     }
 
-    pub fn len(&self) -> usize {
```
Comment (tnowacki) [inline language/move-lang/src/shared/remembering_unique_map.rs:31]: Please revert

File: language/move-lang/src/shared/remembering_unique_map.rs [hunk 16]
```diff
@@ -28,36 +28,15 @@ impl<K: TName, V> RememberingUniqueMap<K, V> {
...
-        self.map.len()
-    }
-
     pub fn add(&mut self, key: K, value: V) -> Result<(), K::Loc> {
         self.map.add(key, value)
     }
 
-    pub fn key_gotten(&self, key: &K) -> bool {
```
Comment (tnowacki) [inline language/move-lang/src/shared/remembering_unique_map.rs:39]: Please revert

File: language/move-lang/src/shared/remembering_unique_map.rs [hunk 17]
```diff
@@ -28,36 +28,15 @@ impl<K: TName, V> RememberingUniqueMap<K, V> {
...
-    pub fn contains_key_(&self, key_: &K::Key) -> bool {
-        self.map.contains_key_(key_)
-    }
-
     pub fn get(&mut self, key: &K) -> Option<&V> {
         self.gotten_keys.insert(key.clone());
         self.map.get(key)
     }
 
-    pub fn get_mut(&mut self, key: &K) -> Option<&mut V> {
```
Comment (tnowacki) [inline language/move-lang/src/shared/remembering_unique_map.rs:56]: Please revert

**Review by tnowacki — CHANGES_REQUESTED** [review]
I started trying to annotate all the spots to revert and what dead code to actually remove. But please don't remove public functions from most of the language dir. (I only saw one so far that can be removed is the `<self>` check in Identifier)

**Comment by huitseeker** [comment]
@tnowacki OK, I can do that — in fact, having that PR in history might serve a historical purpose : those public fns. are definitely **un-tested** and if those are "in place for proper APIs and to ease development in the future" then that's probably what you want to fix . I'll close this in ~a day to give time to others to have a quick glance.

**Comment by tnowacki** [comment]
> those public fns. are definitely un-tested

Most of those public fns are things like `into_inner` or some other field access wrapper. I'm not sure how interesting tests would be for those. 

There is a smaller subsection of public fns are interesting functionality, like grabbing all of the in-edges for a node in the borrow graph. Which should probably be tested

**Comment by dariorussi** [comment]
> I know, I know, @dariorussi doesn't like this one bit.....

LOL.... not one bit.... I mean not even a tiny little teeny teensy bit....
:heart:

**Review by wrwg — CHANGES_REQUESTED** [review]
Appreciate the effort, but please revert all changes on the move prover. Crates like spec_lang provide a framework for multiple tools for language processing we have build today (besides prover, abigen, docgen, errmap) and others we may want to build in the future. I do not have the time right now given the deadline to think hard about what might be needed or not. Also note that in tools, we do not strive for atomic unit testing coverage, but test things as we use them, so no test coverage is rather meaningless for this kind of code.

File: language/borrow-graph/src/graph.rs (hunk 3 above)
Comment (wrwg) [reply language/borrow-graph/src/graph.rs:94]: A note: just last week we absolutely needed this function to work with the borrow graph. We changed representation and not using the borrow graph anymore, but it is an essential functionality for a data structure like this. This is a good example how difficult it is to decide whether to keep such code or not.

**Comment by huitseeker** [comment]
@wrwg Completely fair, as far as I'm concerned: I strive to surface enough signal for you to take good decisions when you can. I'm not strongly opinionated on which side of the fence things should fall on the keep-vs-delete side.

Given what this code is for, there's no surprise that you want to keep a lot of it. I'm more surprised it's untested. 

> Also note that in tools, we do not strive for atomic unit testing coverage, but test things as we use them, so no test coverage is rather meaningless for this kind of code.

One (at least) among tested or used would make me confident the tool is there for a reason. Zero is murky, so I'm showing the ambiguity by this PR. 

The issue IMHO is not that you want to keep that code — that's of course completely fine. It's that you're basing this on contextual thinking you have but that is hard to find in the code base. If I'm wondering if this should be here, there's a good chance another open-source contributor will as well.

**Pull request closed without merging**